### PR TITLE
Core/Spells: Creatures are able to apply diminishing returns.

### LINKED ISSUE #14344: [Scarlet Monastery] NPC Scarlet Gallant - Hammer of Justice
### Current Behaviour

1. Hammer of Justice currently has no DR (Diminishing returns) mechanic.
2. Scarlet Gallant does not cast Hammer of Justice on the current target.

### Expected Blizzlike Behaviour

1. Hammer of Justice should have an DR mechanic
2. Hammer of Justice should also cast on the current target

https://youtu.be/2hKau91Yw6U?t=159 - 2:40, 2:46
Pay attention to the second stun timer, it is only two seconds. Further on this pack there are no stuns at all.

### Source

https://youtu.be/2hKau91Yw6U?t=159 - 2:40, 2:46

### Steps to reproduce the problem

.go cr 40071
You will need a second character for testing because Hammer of Justice is not used on the current target.

### Extra Notes

Original report: https://github.com/chromiecraft/chromiecraft/issues/4530
Maybe this report and the fix to it are related to this problem: https://github.com/azerothcore/azerothcore-wotlk/issues/9028

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/11530ecb95a35f66a2c220781172c57e0749ac48

### Operating system

Windows 10 x64

### Custom changes or Modules

_No response_

### LINKED ISSUE #14348: [Blood Furnace] NPC Nascent Fel Orc - Concussion Blow
### Current Behaviour

1. Concussion Blow currently has no DR (Diminishing returns) mechanic.
2. If the target of Concussion Blow already has this debuff, it is updated instead of adding a new one

### Expected Blizzlike Behaviour

1. Concussion Blow should have an DR mechanic
2. Should be adding a new debuff, not updating/replacing the old one

https://youtu.be/oqOoBDc-ks0?t=451 - 7:32 note, the character receives two debuffs, the duration of one of them is reduced due to the DR mechanic
https://youtu.be/QgYxSueBUBQ?t=1055 - 17:39 analogically

### Source

https://youtu.be/oqOoBDc-ks0?t=451 - 7:32
https://youtu.be/QgYxSueBUBQ?t=1055 - 17:39

### Steps to reproduce the problem

.tele bf
.go xyz 453 52 10

### Extra Notes

Original report: https://github.com/chromiecraft/chromiecraft/issues/4497
Maybe this report and the fix to it are related to this problem: #9028

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/4a2964e10a03a2b6ba2077e6b362b8a6ba6675fb

### Operating system

Windows 10 x64

### Custom changes or Modules

_No response_

### LINKED ISSUE #14350: [Stratholme] Crimson Defender - Hammer of Justice
### Current Behaviour

Hammer of Justice currently has no DR (Diminishing returns) mechanic.

### Expected Blizzlike Behaviour

Hammer of Justice should have an DR mechanic
https://youtu.be/rFp169RCcSs?t=1867 - 31:09 - first Hammer of Justice, 31:22 - second with reduced duration

### Source

https://youtu.be/rFp169RCcSs?t=1867

### Steps to reproduce the problem

.go cr 54140

### Extra Notes

Maybe this report and the fix to it are related to this problem: #9028

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/790f444ffac09bfb9193fbd18aa7c2d730a02600

### Operating system

Windows 10 x64

### Custom changes or Modules

_No response_

### PR DESCRIPTION
**Pull request by UltraNix**
Fixes #14348
Fixes #14344
Fixes #14350

## Issues Addressed:

- Closes #14348
- Parially closes #14344
- Closes #14350
- Closes https://github.com/chromiecraft/chromiecraft/issues/4530
- Closes https://github.com/chromiecraft/chromiecraft/issues/4599
- Closes https://github.com/chromiecraft/chromiecraft/issues/4497

## Tests Performed:

- Tested ingame.

## How to Test the Changes:

`.tele bf`
`.go xyz 453 52 10`

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Nyeriah** [comment]
@Samsequel tested this and some mobs don't have DR. Bosses f.e mostly certainly do not

**Comment by Gultask** [comment]
From what I've seen creatures outside instances also don't have DR. And even inside instances I'm pretty sure some spells don't have DRs, either.

**Comment by UltraNix** [comment]
Not all spells applies diminishing returns.
@Nyeriah @Gultask Some exact examples of spells?

**Comment by Gultask** [comment]
The fear from this NPC does not have DR. Tested it myself
https://www.wowhead.com/npc=19826/dark-conclave-shadowmancer#abilities

> In Hellfire Ramparts now [Classic] and Kidney Shot has DR, Disarm does not, and the knockdown they have doesn't look like it does

Also me

Neifion shared this image in the Discord
image

**Comment by Samsequel** [comment]
The vultures in Hellfire Peninsula that cast swoop should not have DR

**Comment by neifion-00000000** [comment]
> The fear from this NPC does not have DR. Tested it myself https://www.wowhead.com/npc=19826/dark-conclave-shadowmancer#abilities
> 
> > In Hellfire Ramparts now [Classic] and Kidney Shot has DR, Disarm does not, and the knockdown they have doesn't look like it does
> 
> Also me
> 
> Neifion shared this image in the Discord image

yup its important to note that these are NOT the same type of DR the pvp mechanics have

If multiple mobs have the same ability that have DR, the DR should work off the individual creature's use, not the effect on the player.

**Comment by UltraNix** [comment]
> @Samsequel tested this and some mobs don't have DR. Bosses f.e mostly certainly do not e.g King Ymiron

As I can see, bosses are excluded from diminishing returns in purpose.

**Comment by UltraNix** [comment]
> The fear from this NPC does not have DR. Tested it myself https://www.wowhead.com/npc=19826/dark-conclave-shadowmancer#abilities
> 
Tested myself - works okay.

**Comment by UltraNix** [comment]
> > In Hellfire Ramparts now [Classic] and Kidney Shot has DR, Disarm does not, and the knockdown they have doesn't look like it does
> 
> Also me
> 
But what mob?

**Comment by UltraNix** [comment]
> The vultures in Hellfire Peninsula that cast swoop should not have DR

What is the id?

**Comment by Samsequel** [comment]
> > The vultures in Hellfire Peninsula that cast swoop should not have DR
> 
> What is the id?

https://www.wowhead.com/tbc/npc=16972/bonestripper-buzzard

**Comment by UltraNix** [comment]
> 16972

Fixed.

File: src/server/game/Spells/SpellMgr.cpp [hunk 1]
```diff
@@ -81,6 +81,9 @@ DiminishingGroup GetDiminishingReturnsGroupForSpell(SpellInfo const* spellproto,
                 // Screams of the Past
                 else if (spellproto->Id == 7074)
                     return DIMINISHING_NONE;
+                // Swoop
+                else if (spellproto->Id == 37012)
+                    return DIMINISHING_NONE;
```
Comment (Helias) [inline src/server/game/Spells/SpellMgr.cpp:86]: minor: if you have a `return` inside a `if / else if` you could convert the `else if` into `if`.

```C++
                // Pet charge effects (Infernal Awakening, Demon Charge)
                if (spellproto->SpellVisual[0] == 2816 && spellproto->SpellIconID == 15)
                    return DIMINISHING_CONTROLLED_STUN;
                // Gnaw
                if (spellproto->Id == 47481)
                    return DIMINISHING_CONTROLLED_STUN;
                // Screams of the Past
                if (spellproto->Id == 7074)
                    return DIMINISHING_NONE;
                // Swoop
                if (spellproto->Id == 37012)
                    return DIMINISHING_NONE;
```

Comment (UltraNix) [reply src/server/game/Spells/SpellMgr.cpp:86]: I know it. Just don;t want to break codestyle there.

Comment (Helias) [reply src/server/game/Spells/SpellMgr.cpp:86]: ok, no worries at all and thanks for the fix ;-)

**Comment by Gultask** [comment]
We can't possibly go to every spell cast by NPCs to check if it has DR or not. This is a core issue, not something that SpellInfo should or could fix. Making exceptions just seems insane.

**Comment by UltraNix** [comment]
That's why now we have a general mechanic. And now it works- previously none of the mob was allowed to apply diminishing return.

File: src/server/game/Spells/SpellInfoCorrections.cpp [hunk 2]
```diff
@@ -4455,6 +4455,14 @@ void SpellMgr::LoadSpellInfoCorrections()
         spellInfo->Effects[EFFECT_0].TargetB = TARGET_DEST_CASTER;
     });
 
+    // Hammer of Justice
+    // Sonic Burst
+    // Concussion Blow
+    ApplySpellFix({ 5589, 8281,13005, 22427 }, [](SpellInfo* spellInfo)
+    {
+        spellInfo->AttributesEx3 |= SPELL_ATTR3_DOT_STACKING_RULE;
+    });
```
Comment (Kitzunu) [inline src/server/game/Spells/SpellInfoCorrections.cpp:4464]: What does this have to do with DR

Comment (UltraNix) [reply src/server/game/Spells/SpellInfoCorrections.cpp:4464]: Thanks do that you can get multiple stacks of debufs, so every mob can give you its own aura

**Comment by Maelthyrr** [comment]
Tested again. The mechanic work. After I'm not an expert on diminishing returns. And I don't really know which spell mobs should apply or not DR. But in any case, if we have some spells with and others without, we will need something to differentiate them. It's impossible to get a global core mechanic that resolves everything.

File: src/server/game/Spells/SpellInfoCorrections.cpp [hunk 3]
```diff
@@ -4459,6 +4459,18 @@ void SpellMgr::LoadSpellInfoCorrections()
...
+    ApplySpellFix({ 14915 }, [](SpellInfo* spellInfo)
+    {
+        spellInfo->AuraInterruptFlags &= ~AURA_INTERRUPT_FLAG_NOT_SEATED;
+    });
+
+     // Cleansing Totem, Healing Stream Totem, Mana Tide Totem
+    ApplySpellFix({ 8171,52025, 52041, 52042, 52046, 52047, 52048, 52049, 52050, 58759, 58760, 58761, 39610, 39609 }, [](SpellInfo* spellInfo)
+    {
+        spellInfo->AttributesEx2 |= SPELL_ATTR2_IGNORE_LINE_OF_SIGHT;
+    });
```
Comment (Helias) [inline src/server/game/Spells/SpellInfoCorrections.cpp:4472]: can we move these changes to db ideally? (I do not want to block this PR for this, just wondering)

Comment (UltraNix) [reply src/server/game/Spells/SpellInfoCorrections.cpp:4472]: Hm? These come from merge...

Comment (Helias) [reply src/server/game/Spells/SpellInfoCorrections.cpp:4472]: I see, but I was wondering that, because I usually see this ApplySpellFix, but I think that we could use spell_dbc table for it
anyways, it's a bit off-topic

Comment (Nyeriah) [reply src/server/game/Spells/SpellInfoCorrections.cpp:4472]: spell_dbc table should not be used for this kind of change as it obfuscates the custom edits of spells

Comment (Kitzunu) [reply src/server/game/Spells/SpellInfoCorrections.cpp:4472]: spell_dbc is strictly for server-side spells only.

Comment (Helias) [reply src/server/game/Spells/SpellInfoCorrections.cpp:4472]: :ok_hand:

**Review by Kitzunu — APPROVED** [review]

**Comment by Helias** [comment]
@azerothcore/testers

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers please test this PR

**Comment by Nyeriah** [comment]
We’ve discussed this internally thoroughly and we’re not certain this is the proper solution to this

**Comment by neifion-00000000** [comment]
yeah I think this one can be closed unless what I said in my prior comment has been addressed.  The most important thing in my comment that is likely unaddressed is that creature diminishing returns to players are on an individual creature basis, i.e. two creatures using the same moves will have their own DRs running.

General ability improvements have gone in since and I don't believe this problem is occurring much anymore anyway (though, for correctness, at some point someone will want to address it).

**Comment by Jinteki01** [comment]
> yeah I think this one can be closed unless what I said in my prior comment has been addressed. _**The most important thing in my comment that is likely unaddressed is that creature diminishing returns to players are on an individual creature basis, i.e. two creatures using the same moves will have their own DRs running.**_
> 
> General ability improvements have gone in since and I don't believe this problem is occurring much anymore anyway (though, for correctness, at some point someone will want to address it). I also don't know where that mechanic even applies outside certain TBC dungeons, as was said by others/Sam in the thread it is not present in many areas of the game (or if its a universal mechanic, the ability use is infrequent enough to rarely trigger it--even as is it shouldn't happen often).


The bolded section is not the case at all. It does not work like that for players applying stuns to creatures (which is what your original screenshot is stating, not the other way around) and does not work like that for creatures applying stuns to players as well. If there is a warrior who stuns a mob for the full duration then I stun the mob after ward with hammer of justice, my hammer of justice stun only lasts for half the duration. Also if what you are stating is true then the videos I provided in my initial report with separate mobs applying the same spell stuns would have shown the full stun duration. I deliberately timestamped those videos to specifically show different mobs applying a stun with the DR clearly visible. 

This is what makes dungeons like Blood Furnace (one I included in the videos) more of annoyance currently. The mobs (specifically in the gauntlet before the 2nd boss) are applying the full duration stun each time, when that should not be the case.

Also I some mobs were mentioned (the vultures in helffire) as not having DR. This is true because they are not applying a stun, they are doing a 2 second knockdown. The knockdown does not have a DR. This is similar to the ravagers and some other orc mobs who just do a knockdown effect, not a stun.

**Comment by neifion-00000000** [comment]
> 
that might be the case, do we have any proof that categories outside stun are impacted by DR though?  I'm unsure how DRs are handled in the core if they are separated by group properly as well, it looks like this fix was just going to apply PVP DRs to PVE

**Comment by elthehablo** [comment]
> > 
> that might be the case, do we have any proof that categories outside stun are impacted by DR though?  I'm unsure how DRs are handled in the core if they are separated by group properly as well, it looks like this fix was just going to apply PVP DRs to PVE

An issue I heard about is the case of entangling roots. Allegedly it should work less effectively when cast consecutively.
